(LaTeX command is defined in this cell.)
$\newcommand{\bm}[1]{\boldsymbol{#1}}$

# The Matrix

We consider the parity check matrix of an $(n,m,c,\gamma,(1-\epsilon)c)$ lossless bipartite expander with:
1. $n = 40$
2. $m = 19$
3. $c = 3$
4. $\gamma = 5/40 = 0.125$
5. $\epsilon = 0.480$

The matrix $\mathbf{G} = \begin{bmatrix}\mathbf{g} \ \mathbf{\hat{G}}\end{bmatrix} \in \mathbb{Z}^{n \times k}$ where $\mathbf{\hat{G}} = (\mathbf{P}\mathbf{H})^\top$ for some signed permutation matrix $\mathbf{P}$. 

In [1]:
rows = [
    [ 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0],  # r1
    [ 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,-1, 1, 0, 0, 0, 0, 0],  # r2
    [ 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],  # r3
    [ 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-1, 1, 0, 1, 0, 0, 0],  # r4
    [ 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-1,-1],  # r5
    [ 0, 0, 1, 0, 0, 0, 0, 0, 1, 0,-1, 0, 0, 0, 0, 0, 0, 0, 0, 0],  # r6
    [ 1, 0, 0,-1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],  # r7
    [ 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0],  # r8
    [ 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,-1],  # r9
    [ 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,-1],  # r10
    [-1,-1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],  # r11
    [ 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,-1, 0, 0, 0, 0],  # r12
    [ 1, 0, 1,-1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],  # r13
    [ 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,-1, 0, 0, 0, 0, 0,-1],  # r14
    [ 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-1],  # r15
    [-1, 0, 0,-1, 0, 0, 1, 0, 0, 0,-1, 0, 0, 0, 0, 0, 0, 0, 0, 0],  # r16
    [ 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],  # r17
    [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-1, 1, 0, 0, 1, 0, 0, 0, 0, 0],  # r18
    [ 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,-1, 0, 0, 0, 0],  # r19
    [ 1,-1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],  # r20
    [ 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],  # r21
    [ 1,-1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,-1],  # r22
    [ 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],  # r23
    [ 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,-1, 0,-1, 0, 0, 0, 0],  # r24
    [-1, 0, 0,-1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0],  # r25
    [ 1,-1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,-1],  # r26
    [ 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,-1, 0, 0, 0, 0],  # r27
    [ 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0],  # r28
    [ 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,-1, 0],  # r29
    [ 1, 0, 0, 0, 0, 0, 0, 0, 0, 1,-1, 0, 0, 0, 0,-1, 0, 0, 0, 0],  # r30
    [ 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,-1, 0, 0, 0, 0, 0,-1],  # r31
    [ 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,-1, 0, 0, 0, 0, 0, 1, 0, 0, 0],  # r32
    [ 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],  # r33
    [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0,-1, 0],  # r34
    [ 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,-1, 0, 0, 0, 0,-1, 0, 0, 0, 0],  # r35
    [ 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-1, 0],  # r36
    [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,-1, 0],  # r37
    [ 1,-1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-1, 0],  # r38
    [ 1,-1, 0, 0, 0, 0, 0, 0, 0, 0,-1, 0, 0, 0, 0,-1, 0, 0, 0, 0],  # r39
    [ 0, 0, 0,-1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,-1, 0],  # r40
]
G = Matrix(ZZ, rows)

TAG = "G40x20"          # identifies this matrix in UNAUTH_LOG and in table captions

print(TAG, ":", G.nrows(), "x", G.ncols(), "over", G.base_ring())
print("rank:", G.rank())
print("entries in:", sorted(set(ZZ(x) for x in G.list())))
print("max |entry|:", max(abs(ZZ(x)) for x in G.list()))

G40x20 : 40 x 20 over Integer Ring
rank: 20
entries in: [-1, 0, 1]
max |entry|: 1


# Certified Upper Bound on the Reconstruction Coefficient $B_{\bm{\lambda}}(T)$

Certified upper bound on the reconstruction coefficient $B_{\bm{\lambda}}(T)$

**Input.** A matrix $\mathbf{G}\in\mathbb{Z}^{n\times k}$ and a subset $T\subseteq[n]$.

**Output.** The right-hand side of `Lemma:Certified-Coefficient-Bound`,
$$B_{\bm{\lambda}}(T)\;\le\;\Bigl(\|\bm{\lambda}_{T,\mathbb{R}}\|_2^2+R_\mathrm{Babai}(\mathbf{B}_K')^2\Bigr)^{1/2}\;\le\;\Bigl(\tfrac{1}{\sigma_{\min}^+(\mathbf{G}_T)^2}+\tfrac14\textstyle\sum_{i=1}^{d_T}\|\mathbf{b}_i^*\|_2^2\Bigr)^{1/2},$$
together with the measured value $B_{\bm{\lambda}}(T)=\|\bm{\lambda}_T\|_\infty$ for comparison.

The bound has two ingredients, computed separately:

| quantity | meaning | where it comes from |
|---|---|---|
| $\sigma_{\min}^+(\mathbf{G}_T)$ | smallest positive singular value | `Lemma:Real-Least-Norm-Reconstruction` |
| $\|\mathbf{b}_i^*\|_2$ | Gram–Schmidt lengths of $\mathbf{B}_K'$ | Babai nearest-plane radius |

**What to expect.** The bound is loose, for two reasons that are worth separating in the output. The step $\|\cdot\|_\infty\le\|\cdot\|_2$ costs a factor up to $\sqrt{|T|}$, and about $\sqrt{\mathrm{wt}_H(\bm{\lambda}_T)}$ in practice. And $R_\mathrm{Babai}$ bounds the worst corner of the fundamental parallelepiped, whereas the actual decoding usually lands nearer its center. Since $R_\mathrm{Babai}\approx\tfrac{\sqrt{d_T}}{2}D_T$ grows with $d_T=|T|-\mathrm{rank}(\mathbf{G}_T)$, the certificate gets *worse* as $|T|$ grows, while the
measured $B_{\bm{\lambda}}(T)$ gets *better*. The two move in opposite directions; the certificate is a structural result, not a parameter-selection tool.

**Caveat on the kernel basis.** $\mathbf{B}_K'$ must be a basis of the *saturated* kernel $K_T=\ker_\mathbb{Z}(\mathbf{G}_T^\top)$. A basis obtained by clearing denominators from a rational nullspace generally spans a proper finite-index sublattice, which inflates $\det(K_T)$ and hence $R_\mathrm{Babai}$. Sage's `echelonized_basis_matrix()` on `right_kernel(ZZ)` gives the saturated basis; cell 4 checks this via the Smith normal form.

In [2]:
# --- Environment ---
import sys, time, random
from itertools import combinations

from sage.all import (ZZ, QQ, RDF, RR, Matrix, vector, binomial, identity_matrix)
from sage.matrix.special import diagonal_matrix
import pandas as pd

print("sage   :", sys.version.split()[0], "| pandas:", pd.__version__)

sage   : 3.12.11 | pandas: 2.3.2


# The Reconstruction Pipeline (Reused)

The reconstruction pipeline (reused)

These five functions are the canonical pipeline of `Algorithm:Construct-Small-Lambda`, already validated in the smallness experiments. They are needed here only to produce the *measured* value $B_{\bm{\lambda}}(T)=\|\bm{\lambda}_T\|_\infty$, which the certified bound is compared against. They are bundled into one cell because they have been tested separately elsewhere; the new material starts at cell 3.

`construct_lambda` returns `None` when $\mathbf{e}_1^\top\notin\mathrm{Row}_\mathbb{Z}(\mathbf{G}_T)$. The certified bound is defined only for authorized $T$, so callers must check for `None`.

In [3]:
# --- Canonical reconstruction pipeline (validated previously) ---

def linf_norm(v):
    if len(v) == 0:
        return ZZ(0)
    return max(abs(ZZ(x)) for x in v)

def hamming_weight(v):
    return sum(1 for x in v if ZZ(x) != 0)

def hnf_witness(G_S):
    if G_S.nrows() == 0:
        return None, False
    R, U = G_S.hermite_form(transformation=True)
    if R.nrows() == 0:
        return None, False
    e1 = vector(ZZ, [1] + [0]*(G_S.ncols()-1))
    if R.row(0) != e1:
        return None, False
    return vector(ZZ, list(U.row(0))), True

def sample_unique_subsets(n, t, N, seed=0):
    """N distinct random size-t subsets of {0,...,n-1}, as sorted lists."""
    total = int(binomial(n, t))
    if N > total:
        raise ValueError("N=%d exceeds C(%d,%d)=%d" % (N, n, t, total))
    rng = random.Random(int(seed))
    seen, out = set(), []
    while len(out) < N:
        S = tuple(sorted(rng.sample(range(int(n)), int(t))))
        if S not in seen:
            seen.add(S)
            out.append(list(S))
    return out

def kernel_basis(G_S, use_lll=True, delta=0.99):
    """LLL-reduced row-HNF (saturated) basis of ker_Z(G_S^T)."""
    B = G_S.transpose().right_kernel(ZZ).echelonized_basis_matrix()
    if use_lll and B.nrows() > 1:
        B = B.LLL(delta=delta)
    return B

def babai_exact(B, t):
    r, m = B.nrows(), B.ncols()
    if r == 0:
        return vector(ZZ, list(t))
    b = [vector(QQ, list(B.row(i))) for i in range(r)]
    bs = []
    for i in range(r):
        v = b[i]
        for j in range(i):
            d = bs[j].dot_product(bs[j])
            if d != 0:
                v = v - (b[i].dot_product(bs[j])/d) * bs[j]
        bs.append(v)
    w = vector(QQ, list(t))
    for i in reversed(range(r)):
        d = bs[i].dot_product(bs[i])
        if d == 0:
            continue
        ch = (w.dot_product(bs[i])/d + QQ(1)/2).floor()
        if ch != 0:
            w = w - QQ(ch)*b[i]
    return vector(ZZ, [ZZ(x) for x in w])

def greedy_reduce_linf(lam0, B, passes=2, window=2):
    lam  = vector(ZZ, list(lam0))
    best = linf_norm(lam)
    if B.nrows() == 0:
        return lam
    for _ in range(int(passes)):
        improved = False
        for i in range(B.nrows()):
            bi = vector(ZZ, list(B.row(i)))
            j  = max(range(len(lam)), key=lambda t: abs(lam[t]))
            if bi[j] == 0:
                continue
            a0 = (QQ(-lam[j])/QQ(bi[j]) + QQ(1)/2).floor()
            lam_best, norm_best = lam, best
            for a in range(int(a0-window), int(a0+window)+1):
                cand = lam + ZZ(a)*bi
                cn   = linf_norm(cand)
                if cn < norm_best:
                    lam_best, norm_best = cand, cn
            if norm_best < best:
                lam, best, improved = lam_best, norm_best, True
        if not improved:
            break
    return lam

def construct_lambda(G, S, passes=2, window=2):
    """Canonical reconstruction vector, or None if S is unauthorized."""
    S = list(S)
    G_S = G.matrix_from_rows(S)
    lam0, ok = hnf_witness(G_S)
    if not ok:
        return None
    B = kernel_basis(G_S)
    if B.nrows() == 0:
        return lam0
    return greedy_reduce_linf(babai_exact(B, lam0), B, passes=passes, window=window)

print("pipeline loaded")

pipeline loaded


# The First Term: $\|\bm{\lambda}_{T,\mathbb{R}}\|_2 \le 1/\sigma_{\min}^+(\mathbf{G}_T)$

The first term: $\|\bm{\lambda}_{T,\mathbb{R}}\|_2 \le 1/\sigma_{\min}^+(\mathbf{G}_T)$

By `Lemma:Real-Least-Norm-Reconstruction`, the least-$\ell_2$-norm *real* solution of $\mathbf{A}_T\bm{\lambda}=\mathbf{e}_1$ satisfies $\|\bm{\lambda}_{T,\mathbb{R}}\|_2 \le 1/\sigma_{\min}^+(\mathbf{G}_T)$, where $\sigma_{\min}^+$ is the smallest *positive* singular value. This term is the part of the coefficient size that no amount of lattice reduction can remove: it is the distance from the origin to the affine solution space, and every integer solution is at least this long.

**Why "positive" needs care.** $\mathbf{G}_T$ is generally rank-deficient - a right-vertex of the expander whose neighbors all lie outside $T$ produces a zero column - so the smallest singular value is $0$ and must be discarded. Selecting by a magnitude threshold (`sv > 1e-9`) is fragile: a genuinely small non-zero singular value could be dropped, or
numerical noise kept. Instead we compute the rank $r_T$ *exactly* over $\mathbb{Z}$ with `G_S.rank()`, then take the $r_T$-th largest singular value. The rank is exact; only the singular value itself is numerical.

We also return $\|\bm{\lambda}_{T,\mathbb{R}}\|_2$ computed directly via the pseudoinverse, since `Lemma:Real-Least-Norm-Reconstruction` gives the bound $1/\sigma_{\min}^+$ only as an upper estimate — the exact value is usually smaller, and reporting both shows how much slack this step alone contributes.

In [4]:
def real_least_norm(G_S, verbose=False):
    """First term of the certified bound.

    Returns a dict with:
      rank        exact rank of G_S over ZZ
      d_T         |T| - rank  (dimension of the real kernel)
      sigma_min   smallest POSITIVE singular value, selected by exact rank
      bound       1 / sigma_min        (the bound used in the lemma)
      exact       ||lambda_{T,R}||_2   (via pseudoinverse; <= bound)

    The SVD is computed over RDF.  
    On an integer matrix, singular_values() is exact and goes through the characteristic polynomial, which is far too slow at these sizes; 
    RDF uses a numerical LAPACK routine instead.
    The rank is still taken exactly over ZZ, so the choice of which singular value is "smallest positive" involves no threshold.
    """
    m  = G_S.nrows()
    rT = G_S.rank()                                 # exact, over ZZ
    if rT == 0:
        return None

    A_rdf = G_S.change_ring(RDF)                    # numerical
    sv = sorted((float(x) for x in A_rdf.singular_values()), reverse=True)
    sigma_min = sv[rT - 1]                          # rT-th largest = smallest positive

    # least-norm real solution: lambda_{T,R} = (G_T^T)^+ e_1
    A     = A_rdf.transpose()                       # k x |T|
    e1    = vector(RDF, [1] + [0]*(G_S.ncols()-1))
    lam_R = A.pseudoinverse() * e1
    exact = float(lam_R.norm(2))

    out = dict(rank=rT, d_T=m - rT, sigma_min=sigma_min,
               bound=1.0/sigma_min, exact=exact)
    if verbose:
        print("  rank=%d  d_T=%d  sigma_min=%.6f  1/sigma_min=%.4f  ||lam_R||_2=%.4f"
              % (out['rank'], out['d_T'], out['sigma_min'], out['bound'], out['exact']))
        print("  singular values (largest 5):", [round(x, 4) for x in sv[:5]])
        print("  singular values (around cut):",
              [round(x, 6) for x in sv[max(0, rT-2):rT+2]])
    return out

Self-check.

In [5]:
# --- self-check ---
_S = list(range(26))
_G_S = G.matrix_from_rows(_S)
_lam, _ok = hnf_witness(_G_S)
print("authorized:", _ok)
_r = real_least_norm(_G_S, verbose=True)

# the exact least-norm solution must satisfy the bound and the equation
assert _r['exact'] <= _r['bound'] + 1e-9
assert _r['rank'] + _r['d_T'] == _G_S.nrows()
print("\nsanity: ||lam_R||_2 <= 1/sigma_min  OK")

# and any integer reconstruction vector is at least as long
if _ok:
    _li = construct_lambda(G, _S)
    print("||lambda_T||_2 =", float(vector(RDF, list(_li)).norm(2)),
          " >= ||lam_R||_2 =", round(_r['exact'], 4))
    assert float(vector(RDF, list(_li)).norm(2)) >= _r['exact'] - 1e-9

authorized: True
  rank=20  d_T=6  sigma_min=0.588595  1/sigma_min=1.6990  ||lam_R||_2=0.4481
  singular values (largest 5): [5.1739, 3.2196, 2.7889, 2.701, 2.56]
  singular values (around cut): [0.697835, 0.588595]

sanity: ||lam_R||_2 <= 1/sigma_min  OK
||lambda_T||_2 = 2.8284271247461903  >= ||lam_R||_2 = 0.4481


# The Second Term: $R_\mathrm{Babai}(\mathbf{B}_K') = \tfrac12\bigl(\sum_i\|\mathbf{b}_i^*\|_2^2\bigr)^{1/2}$

The second term: $R_\mathrm{Babai}(\mathbf{B}_K') = \tfrac12\bigl(\sum_i\|\mathbf{b}_i^*\|_2^2\bigr)^{1/2}$

Babai's nearest-plane algorithm returns a point whose in-span error lies in the fundamental region $\{\sum_i x_i\mathbf{b}_i^*: x_i\in[-\tfrac12,\tfrac12)\}$. The worst corner of that region is at distance $\tfrac12(\sum_i\|\mathbf{b}_i^*\|_2^2)^{1/2}$ from the origin, which is the standard radius.

Two things are worth computing alongside it.

**Saturation.** $R_\mathrm{Babai}$ is meaningful only if $\mathbf{B}_K'$ spans the full $K_T=\ker_\mathbb{Z}(\mathbf{G}_T^\top)$ rather than a finite-index sublattice. A sublattice has larger covolume and longer Gram–Schmidt vectors, inflating the radius. We verify saturation via the Smith normal form: all elementary divisors must be $1$.

**The realized error.** $R_\mathrm{Babai}$ is a worst-corner bound; the actual decoding error $R_\mathrm{dec}(T)=\|\mathbf{t}_T-\bm{\nu}_\mathrm{dec}\|_2$ is typically much smaller, since a random point of the region sits nearer its center. Reporting both separates "how loose is the Babai radius" from "how loose is $\ell_2\to\ell_\infty$".

In [6]:
from sage.matrix.matrix_space import MatrixSpace

def babai_radius(G_S, verbose=False):
    """Second term of the certified bound, plus saturation and realised-error diagnostics."""
    B = kernel_basis(G_S)                      # LLL-reduced saturated HNF basis
    r = B.nrows()
    if r == 0:
        return dict(d_T=0, R_Babai=0.0, gs_sq=[], saturated=True,
                    det_KT=1.0, D_T=None, R_dec=0.0)

    b, bs = [vector(QQ, list(B.row(i))) for i in range(r)], []
    for i in range(r):
        v = b[i]
        for j in range(i):
            d = bs[j].dot_product(bs[j])
            if d != 0:
                v = v - (b[i].dot_product(bs[j])/d) * bs[j]
        bs.append(v)
    gs_sq  = [QQ(u.dot_product(u)) for u in bs]
    R_bab  = float(RDF(sum(gs_sq)).sqrt() / 2)

    # saturation: all Smith elementary divisors must be 1
    divs = [ZZ(x) for x in B.smith_form()[0].diagonal() if x != 0]
    saturated = all(abs(d) == 1 for d in divs)

    det_KT = float(RDF(prod(gs_sq)).sqrt())
    D_T    = det_KT ** (1.0/r)

    # realised decoding error: ||lambda_dec - lambda_{T,R}||_2
    lam0, ok = hnf_witness(G_S)
    R_dec = None
    if ok:
        lam_dec = babai_exact(B, lam0)
        A     = G_S.transpose().change_ring(RDF)
        e1    = vector(RDF, [1] + [0]*(G_S.ncols()-1))
        lam_R = A.pseudoinverse() * e1
        R_dec = float((vector(RDF, list(lam_dec)) - lam_R).norm(2))

    out = dict(d_T=r, R_Babai=R_bab, gs_sq=gs_sq, saturated=saturated,
               divisors=divs, det_KT=det_KT, D_T=D_T, R_dec=R_dec)
    if verbose:
        print("  d_T=%d  saturated=%s  det(K_T)=%.3f  D_T=%.3f"
              % (r, saturated, det_KT, D_T))
        print("  ||b_i*||_2:", [round(float(RDF(x).sqrt()), 3) for x in gs_sq])
        print("  R_Babai=%.4f   realised R_dec=%.4f   ratio=%.2f"
              % (R_bab, R_dec, R_bab/R_dec if R_dec else float('nan')))
    return out

Self-check.

In [7]:
# --- self-check ---
_bb = babai_radius(_G_S, verbose=True)
assert _bb['saturated'], "kernel basis is NOT saturated: %s" % (_bb['divisors'],)
assert _bb['d_T'] == _r['d_T']
assert _bb['R_dec'] <= _bb['R_Babai'] + 1e-9, "realised error exceeds the Babai radius"
print("\nsanity: R_dec <= R_Babai, basis saturated  OK")

  d_T=6  saturated=True  det(K_T)=8863.469  D_T=4.549
  ||b_i*||_2: [4.0, 4.123, 5.026, 4.725, 4.665, 4.852]
  R_Babai=5.6099   realised R_dec=2.7927   ratio=2.01

sanity: R_dec <= R_Babai, basis saturated  OK


# Assembling the Bound

Assembling the bound

`certified_bound(G, T)` combines the two terms and returns them alongside the measured value, so the slack in each step is visible. It returns `None` for unauthorized $T$, since
$B_{\bm{\lambda}}(T)$ is defined only on $\Gamma_1$.

Three ratios are reported:

- `slack_sigma` $= (1/\sigma_{\min}^+)\,/\,\|\bm{\lambda}_{T,\mathbb{R}}\|_2$ — how much `Lemma:Real-Least-Norm-Reconstruction` gives away by assuming $\mathbf{e}_1$ aligns with the
  worst singular direction;
- `slack_babai` $= R_\mathrm{Babai}/R_\mathrm{dec}$ — worst corner versus realized decoding;
- `slack_norm` $= \|\bm{\lambda}_T\|_2/\|\bm{\lambda}_T\|_\infty$ — the $\ell_2\to\ell_\infty$ transition, at most $\sqrt{\mathrm{wt}_H(\bm{\lambda}_T)}$.

The product of the three is roughly the overall ratio `certified / measured`.

In [8]:
def certified_bound(G, T, passes=2, window=2, verbose=False):
    """Certified upper bound on B_lambda(T), with the measured value for comparison."""
    T = list(T)
    G_S = G.matrix_from_rows(T)
    lam0, ok = hnf_witness(G_S)
    if not ok:
        return None                                  # T unauthorized

    rl = real_least_norm(G_S)
    bb = babai_radius(G_S)

    tight = float(RDF(rl['exact']**2 + (bb['R_dec'] or 0)**2).sqrt())
    cert  = float(RDF(rl['bound']**2 + bb['R_Babai']**2).sqrt())

    lam_T = construct_lambda(G, T, passes=passes, window=window)
    meas_inf = int(linf_norm(lam_T))
    meas_2   = float(vector(RDF, list(lam_T)).norm(2))

    out = dict(size=len(T), d_T=bb['d_T'], sigma_min=rl['sigma_min'],
               lam_R=rl['exact'], sigma_bound=rl['bound'],
               R_dec=bb['R_dec'], R_Babai=bb['R_Babai'], D_T=bb['D_T'],
               tight=tight, certified=cert,
               measured_inf=meas_inf, measured_2=meas_2,
               wt=int(hamming_weight(lam_T)),
               slack_sigma=rl['bound']/rl['exact'],
               slack_babai=bb['R_Babai']/bb['R_dec'] if bb['R_dec'] else None,
               slack_norm=meas_2/meas_inf if meas_inf else None,
               ratio=cert/meas_inf if meas_inf else None)
    if verbose:
        print("  |T|=%d  d_T=%d  D_T=%.3f" % (out['size'], out['d_T'], out['D_T']))
        print("  term 1: ||lam_R||_2=%.4f  <=  1/sigma_min=%.4f   (slack %.2f)"
              % (out['lam_R'], out['sigma_bound'], out['slack_sigma']))
        print("  term 2: R_dec=%.4f       <=  R_Babai=%.4f       (slack %.2f)"
              % (out['R_dec'], out['R_Babai'], out['slack_babai']))
        print("  bound : realised=%.4f    <=  certified=%.4f" % (out['tight'], out['certified']))
        print("  actual: ||lam_T||_inf=%d  ||lam_T||_2=%.4f  wt=%d  (slack %.2f)"
              % (out['measured_inf'], out['measured_2'], out['wt'], out['slack_norm']))
        print("  certified / measured = %.2f" % out['ratio'])
    return out

Self-check.

In [9]:
# --- self-check ---
_c = certified_bound(G, list(range(26)), verbose=True)
assert _c['measured_inf'] <= _c['certified'] + 1e-9, "BOUND VIOLATED"
assert _c['tight'] <= _c['certified'] + 1e-9
print("\nsanity: measured <= certified  OK")
assert certified_bound(G, [0, 1, 2]) is None or True   # small T may be unauthorized

  |T|=26  d_T=6  D_T=4.549
  term 1: ||lam_R||_2=0.4481  <=  1/sigma_min=1.6990   (slack 3.79)
  term 2: R_dec=2.7927       <=  R_Babai=5.6099       (slack 2.01)
  bound : realised=2.8284    <=  certified=5.8615
  actual: ||lam_T||_inf=1  ||lam_T||_2=2.8284  wt=8  (slack 2.83)
  certified / measured = 5.86

sanity: measured <= certified  OK


Self-test, single subset.

In [10]:
%%time
_r = real_least_norm(_G_S, verbose=True)
_c = certified_bound(G, list(range(26)), verbose=True)

  rank=20  d_T=6  sigma_min=0.588595  1/sigma_min=1.6990  ||lam_R||_2=0.4481
  singular values (largest 5): [5.1739, 3.2196, 2.7889, 2.701, 2.56]
  singular values (around cut): [0.697835, 0.588595]
  |T|=26  d_T=6  D_T=4.549
  term 1: ||lam_R||_2=0.4481  <=  1/sigma_min=1.6990   (slack 3.79)
  term 2: R_dec=2.7927       <=  R_Babai=5.6099       (slack 2.01)
  bound : realised=2.8284    <=  certified=5.8615
  actual: ||lam_T||_inf=1  ||lam_T||_2=2.8284  wt=8  (slack 2.83)
  certified / measured = 5.86
CPU times: user 11.2 ms, sys: 0 ns, total: 11.2 ms
Wall time: 10.5 ms


# Sweeping Subsets and Sizes

Sweeping subsets and sizes

The single-subset check confirms the lemma holds; the interesting question is how the bound behaves across $T$. `bound_curve` samples subsets per size and tabulates the two terms, the certified bound, and the measured value.

Two predictions to test:

**The bound should worsen as $|T|$ grows.** $R_\mathrm{Babai}\approx\tfrac{\sqrt{d_T}}{2}D_T$ and $d_T = |T|-\mathrm{rank}(\mathbf{G}_T)$ increases with $|T|$, while the measured
$\beta_{\bm{\lambda}}(s)$ decreases (4 at $s=26$ down to 1 at $s=40$). So `ratio` should grow with $s$ — the certificate is loosest exactly where the truth is tightest.

**The balanced-profile heuristic should hold.** With a flat Gram–Schmidt profile, $R_\mathrm{Babai}\approx\tfrac{\sqrt{d_T}}{2}D_T$. At $|T|=26$ this gave $5.57$ against an
actual $5.61$. The `heur_ratio` column tracks this across sizes and validates the "Heuristic Decoding-Scale Estimate" subsection empirically.

In [11]:
def bound_curve(G, sizes, N=20, seed=0, passes=2, window=2, verbose=True):
    """Certified bound vs measured value across subset sizes.

    N subsets per size, or exhaustive enumeration when C(n,s) <= N.
    """
    n, rnd, recs = G.nrows(), random.Random(int(seed)), []
    for s in sizes:
        t0  = time.time()
        tot = int(binomial(n, s))
        if tot <= N:
            subs = list(combinations(range(n), s))
            mode = "exact"
        else:
            seen = set()
            while len(seen) < N:
                seen.add(tuple(sorted(rnd.sample(range(int(n)), int(s)))))
            subs, mode = sorted(seen), "mc"

        if verbose:
            print("s=%2d [%5s] %d subsets ..." % (s, mode, len(subs)), end=" ", flush=True)

        rows_s = []
        for T in subs:
            c = certified_bound(G, list(T), passes=passes, window=window)
            if c is not None:
                rows_s.append(c)
        if not rows_s:
            if verbose:
                print("all unauthorized")
            continue

        cert = sorted(r['certified'] for r in rows_s)
        rat  = sorted(r['ratio']     for r in rows_s)
        agg = dict(
            s=s, mode=mode, trials=len(rows_s), unauth=len(subs)-len(rows_s),
            d_T=int(sum(r['d_T'] for r in rows_s)/len(rows_s)),
            D_T=round(sum(r['D_T'] for r in rows_s)/len(rows_s), 3),
            R_Babai=round(sum(r['R_Babai'] for r in rows_s)/len(rows_s), 3),
            heur=round(sum((r['d_T']**0.5/2)*r['D_T'] for r in rows_s)/len(rows_s), 3),
            cert_med=round(cert[len(cert)//2], 2),
            cert_max=round(cert[-1], 2),
            meas_max=max(r['measured_inf'] for r in rows_s),
            ratio_med=round(rat[len(rat)//2], 2),
            slack_sigma=round(sum(r['slack_sigma'] for r in rows_s)/len(rows_s), 2),
            slack_babai=round(sum(r['slack_babai'] for r in rows_s)/len(rows_s), 2),
            slack_norm=round(sum(r['slack_norm'] for r in rows_s)/len(rows_s), 2),
            secs=round(time.time()-t0, 1))
        agg['heur_ratio'] = round(agg['R_Babai']/agg['heur'], 3)
        recs.append(agg)
        if verbose:
            print("d_T=%2d D_T=%.3f R_Babai=%.3f (heur %.3f, %.3f) "
                  "cert med/max=%.2f/%.2f meas_max=%d ratio=%.2f  %.1fs"
                  % (agg['d_T'], agg['D_T'], agg['R_Babai'], agg['heur'],
                     agg['heur_ratio'], agg['cert_med'], agg['cert_max'],
                     agg['meas_max'], agg['ratio_med'], agg['secs']))
    return pd.DataFrame(recs).set_index('s')

Sweeping test.

In [12]:
%%time
df_bound = bound_curve(G, sizes=range(26, 41), N=200, seed=0)
display(df_bound)

s=26 [   mc] 200 subsets ... d_T= 6 D_T=4.375 R_Babai=5.606 (heur 5.383, 1.041) cert med/max=6.09/7.82 meas_max=3 ratio=5.96  1.5s
s=27 [   mc] 200 subsets ... d_T= 7 D_T=3.927 R_Babai=5.326 (heur 5.197, 1.025) cert med/max=5.77/7.17 meas_max=2 ratio=5.72  1.7s
s=28 [   mc] 200 subsets ... d_T= 8 D_T=3.572 R_Babai=5.126 (heur 5.051, 1.015) cert med/max=5.48/6.41 meas_max=2 ratio=5.48  1.8s
s=29 [   mc] 200 subsets ... d_T= 9 D_T=3.322 R_Babai=5.031 (heur 4.983, 1.010) cert med/max=5.32/5.77 meas_max=2 ratio=5.32  2.1s
s=30 [   mc] 200 subsets ... d_T=10 D_T=3.115 R_Babai=4.956 (heur 4.926, 1.006) cert med/max=5.21/5.54 meas_max=1 ratio=5.21  2.5s
s=31 [   mc] 200 subsets ... d_T=11 D_T=2.950 R_Babai=4.918 (heur 4.891, 1.006) cert med/max=5.15/5.42 meas_max=1 ratio=5.15  2.7s
s=32 [   mc] 200 subsets ... d_T=12 D_T=2.813 R_Babai=4.902 (heur 4.872, 1.006) cert med/max=5.10/5.34 meas_max=1 ratio=5.10  2.9s
s=33 [   mc] 200 subsets ... d_T=13 D_T=2.692 R_Babai=4.886 (heur 4.854, 1.007) cer

,mode,trials,unauth,d_T,D_T,R_Babai,heur,cert_med,cert_max,meas_max,ratio_med,slack_sigma,slack_babai,slack_norm,secs,heur_ratio
s,,,,,,,,,,,,,,,,
26,mc,200,0,6,4.375,5.606,5.383,6.09,7.82,3,5.96,4.04,1.86,2.85,1.5,1.041
27,mc,200,0,7,3.927,5.326,5.197,5.77,7.17,2,5.72,3.9,1.81,2.96,1.7,1.025
28,mc,199,1,8,3.572,5.126,5.051,5.48,6.41,2,5.48,3.79,1.8,2.94,1.8,1.015
29,mc,200,0,9,3.322,5.031,4.983,5.32,5.77,2,5.32,3.58,1.81,2.86,2.1,1.01
30,mc,200,0,10,3.115,4.956,4.926,5.21,5.54,1,5.21,3.52,1.81,2.83,2.5,1.006
31,mc,200,0,11,2.95,4.918,4.891,5.15,5.42,1,5.15,3.33,1.84,2.76,2.7,1.006
32,mc,200,0,12,2.813,4.902,4.872,5.1,5.34,1,5.1,3.23,1.84,2.75,2.9,1.006
33,mc,200,0,13,2.692,4.886,4.854,5.06,5.21,1,5.06,3.2,1.85,2.71,3.3,1.007
34,mc,200,0,14,2.591,4.895,4.848,5.05,5.23,1,5.05,3.09,1.85,2.71,3.6,1.01


CPU times: user 42.9 s, sys: 107 ms, total: 43 s
Wall time: 43 s


# Further Experiment

Fused per-subset computation.

In [14]:
def bound_and_lambda(G, T, passes=2, window=2):
    """Certified bound and measured lambda_T in one pass. None if unauthorized."""
    T = list(T)
    G_S = G.matrix_from_rows(T)

    lam0, ok = hnf_witness(G_S)                       # one HNF
    if not ok:
        return None

    B = kernel_basis(G_S)                             # one kernel + LLL
    r = B.nrows()

    # ---- one Gram-Schmidt, reused for R_Babai and for Babai decoding ----
    if r > 0:
        b, bs = [vector(QQ, list(B.row(i))) for i in range(r)], []
        for i in range(r):
            v = b[i]
            for j in range(i):
                d = bs[j].dot_product(bs[j])
                if d != 0:
                    v = v - (b[i].dot_product(bs[j])/d) * bs[j]
            bs.append(v)
        gs_sq  = [u.dot_product(u) for u in bs]
        R_bab  = float(RDF(sum(gs_sq)).sqrt() / 2)
        D_T    = float(RDF(prod(gs_sq)).sqrt()) ** (1.0/r)

        w = vector(QQ, list(lam0))                    # Babai, reusing bs
        for i in reversed(range(r)):
            d = gs_sq[i]
            if d == 0:
                continue
            ch = (w.dot_product(bs[i])/d + QQ(1)/2).floor()
            if ch != 0:
                w = w - QQ(ch)*b[i]
        lam_dec = vector(ZZ, [ZZ(x) for x in w])
        lam_T   = greedy_reduce_linf(lam_dec, B, passes=passes, window=window)
    else:
        R_bab, D_T, lam_dec, lam_T = 0.0, None, lam0, lam0

    # ---- first term ----
    rT    = G_S.rank()
    A_rdf = G_S.change_ring(RDF)
    sv    = sorted((float(x) for x in A_rdf.singular_values()), reverse=True)
    smin  = sv[rT-1]
    e1    = vector(RDF, [1] + [0]*(G_S.ncols()-1))
    lam_R = A_rdf.transpose().pseudoinverse() * e1

    cert  = float(RDF(1.0/smin**2 + R_bab**2).sqrt())
    R_dec = float((vector(RDF, list(lam_dec)) - lam_R).norm(2))
    tight = float(RDF(float(lam_R.norm(2))**2 + R_dec**2).sqrt())

    return dict(d_T=r, D_T=D_T, sigma_min=smin, lam_R=float(lam_R.norm(2)),
                R_Babai=R_bab, R_dec=R_dec, certified=cert, tight=tight,
                linf=int(linf_norm(lam_T)), wt=int(hamming_weight(lam_T)),
                l2=float(vector(RDF, list(lam_T)).norm(2)))

Curve, mirorring `freq_curve'.

In [15]:
def bound_table_for_s(G, s, tag, N=None, seed=0, passes=2, window=2, verbose=True):
    n, tot = G.nrows(), int(binomial(G.nrows(), s))
    if N is None or N >= tot:
        subs, mode, target = combinations(range(n), s), "exact", tot
    else:
        subs, mode, target = sample_unique_subsets(n, s, int(N), seed), "mc", int(N)

    certs, n_auth, n_unauth = [], 0, 0
    linf_max, cert_at_argmax, argmax_T = 0, None, None
    heur_rat, sl_sig, sl_bab, sl_nrm = [], [], [], []

    t0 = time.time()
    for S in subs:
        d = bound_and_lambda(G, S, passes, window)
        if d is None:
            n_unauth += 1
            continue
        n_auth += 1
        certs.append(d['certified'])
        if d['linf'] > linf_max:                       # track the argmax subset
            linf_max, cert_at_argmax, argmax_T = d['linf'], d['certified'], list(S)
        if d['d_T'] > 0:
            heur_rat.append(d['R_Babai'] / ((d['d_T']**0.5/2) * d['D_T']))
            sl_sig.append((1.0/d['sigma_min']) / d['lam_R'])
            if d['R_dec'] > 0:
                sl_bab.append(d['R_Babai'] / d['R_dec'])
        if d['linf'] > 0:
            sl_nrm.append(d['l2'] / d['linf'])
    dt = time.time() - t0
    assert n_auth + n_unauth == target

    certs.sort()
    mean = lambda v: round(sum(v)/len(v), 3) if v else None
    out = dict(s=s, mode=mode, tested=target, authorized=n_auth, unauthorized=n_unauth,
               linf_max=linf_max, cert_at_argmax=cert_at_argmax, argmax_T=argmax_T,
               cert_min=certs[0], cert_med=certs[len(certs)//2], cert_max=certs[-1],
               heur_ratio=mean(heur_rat), slack_sigma=mean(sl_sig),
               slack_babai=mean(sl_bab), slack_norm=mean(sl_nrm), secs=dt)
    if verbose:
        print("s=%2d [%5s] auth=%7d unauth=%5d | linf_max=%d cert@argmax=%.2f | "
              "cert min/med/max=%.2f/%.2f/%.2f | heur=%.3f | %6.1fs"
              % (s, mode, n_auth, n_unauth, linf_max, cert_at_argmax,
                 out['cert_min'], out['cert_med'], out['cert_max'],
                 out['heur_ratio'], dt))
    return out


def bound_curve_full(G, tag, s_range, N=None, seed=0, repeats=1,
                     passes=2, window=2, verbose=True):
    """Same seeding and MC/exhaustive convention as freq_curve."""
    out = []
    for s in s_range:
        tot   = int(binomial(G.nrows(), s))
        n_rep = 1 if (N is None or N >= tot) else int(repeats)
        merged = None
        for r in range(n_rep):
            res = bound_table_for_s(G, s, tag, N=N, seed=int(seed)+1000*int(s)+r,
                                    passes=passes, window=window,
                                    verbose=(verbose and n_rep == 1))
            if merged is None:
                merged = res
            else:
                for k in ('tested', 'authorized', 'unauthorized', 'secs'):
                    merged[k] += res[k]
                merged['cert_min'] = min(merged['cert_min'], res['cert_min'])
                merged['cert_max'] = max(merged['cert_max'], res['cert_max'])
                merged['cert_med'] = (merged['cert_med'] + res['cert_med'])/2
                if res['linf_max'] > merged['linf_max']:
                    merged['linf_max']      = res['linf_max']
                    merged['cert_at_argmax'] = res['cert_at_argmax']
                    merged['argmax_T']       = res['argmax_T']
                for k in ('heur_ratio', 'slack_sigma', 'slack_babai', 'slack_norm'):
                    merged[k] = round((merged[k] + res[k])/2, 3)
        merged['repeats'] = n_rep
        if verbose and n_rep > 1:
            print("s=%2d [%5s x%d] auth=%7d unauth=%5d | linf_max=%d cert@argmax=%.2f | "
                  "cert min/med/max=%.2f/%.2f/%.2f | heur=%.3f | %6.1fs"
                  % (s, merged['mode'], n_rep, merged['authorized'], merged['unauthorized'],
                     merged['linf_max'], merged['cert_at_argmax'], merged['cert_min'],
                     merged['cert_med'], merged['cert_max'], merged['heur_ratio'],
                     merged['secs']))
        out.append(merged)
    return out

Small test.

In [16]:
%%time
_ = bound_table_for_s(G, 26, TAG, N=2000, seed=0)

s=26 [   mc] auth=   1991 unauth=    9 | linf_max=3 cert@argmax=7.82 | cert min/med/max=5.07/6.09/9.47 | heur=1.040 |    6.9s
CPU times: user 6.87 s, sys: 28 ms, total: 6.9 s
Wall time: 6.9 s


In [17]:
%%time
bc_40 = bound_curve_full(G, TAG, range(26, 41), N=500, seed=0, repeats=1)

s=26 [   mc] auth=    495 unauth=    5 | linf_max=3 cert@argmax=8.73 | cert min/med/max=5.14/6.10/8.73 | heur=1.039 |    1.6s
s=27 [   mc] auth=    496 unauth=    4 | linf_max=2 cert@argmax=5.96 | cert min/med/max=5.03/5.74/7.10 | heur=1.024 |    1.8s
s=28 [   mc] auth=    500 unauth=    0 | linf_max=2 cert@argmax=5.64 | cert min/med/max=4.88/5.50/6.22 | heur=1.015 |    2.0s
s=29 [   mc] auth=    500 unauth=    0 | linf_max=2 cert@argmax=5.36 | cert min/med/max=4.94/5.33/5.83 | heur=1.009 |    2.3s
s=30 [   mc] auth=    500 unauth=    0 | linf_max=2 cert@argmax=5.39 | cert min/med/max=4.86/5.23/5.68 | heur=1.007 |    2.6s
s=31 [   mc] auth=    500 unauth=    0 | linf_max=1 cert@argmax=5.00 | cert min/med/max=4.80/5.16/5.44 | heur=1.005 |    2.8s
s=32 [   mc] auth=    500 unauth=    0 | linf_max=1 cert@argmax=4.95 | cert min/med/max=4.84/5.09/5.34 | heur=1.006 |    3.0s
s=33 [   mc] auth=    500 unauth=    0 | linf_max=1 cert@argmax=4.94 | cert min/med/max=4.87/5.06/5.26 | heur=1.007 | 

In [18]:
df40 = pd.DataFrame([{k: v for k, v in d.items() if k != 'argmax_T'}
                     for d in bc_40]).set_index('s')

In [19]:
display(df40)

,mode,tested,authorized,unauthorized,linf_max,cert_at_argmax,cert_min,cert_med,cert_max,heur_ratio,slack_sigma,slack_babai,slack_norm,secs,repeats
s,,,,,,,,,,,,,,,
26,mc,500,495,5,3,8.731463,5.135931,6.097978,8.731463,1.039,4.115,1.806,2.944,1.563309,1
27,mc,500,496,4,2,5.961273,5.031105,5.741687,7.100020,1.024,3.904,1.794,2.965,1.839985,1
28,mc,500,500,0,2,5.644903,4.880887,5.503313,6.222088,1.015,3.802,1.818,2.916,1.951021,1
29,mc,500,500,0,2,5.356475,4.943174,5.328902,5.826671,1.009,3.609,1.811,2.859,2.316599,1
30,mc,500,500,0,2,5.394082,4.855004,5.228707,5.681540,1.007,3.539,1.801,2.827,2.580464,1
31,mc,500,500,0,1,5.000676,4.800063,5.157383,5.443015,1.005,3.384,1.822,2.786,2.756186,1
32,mc,500,500,0,1,4.947458,4.844025,5.090944,5.340176,1.006,3.279,1.828,2.755,2.955470,1
33,mc,500,500,0,1,4.942510,4.865909,5.059311,5.258872,1.007,3.142,1.822,2.752,3.109639,1
34,mc,500,500,0,1,5.047143,4.881743,5.048843,5.212823,1.009,3.113,1.831,2.738,3.406192,1
